discovery/registry.go

```diff
@@ -0,0 +1,261 @@
+	// This function can be removed if https://github.com/go-yaml/yaml/issues/642 is fixed.
+
+	outVal := reflect.ValueOf(out)
+	if outVal.Kind() != reflect.Ptr {
+		return fmt.Errorf("discovery: can only unmarshal into a struct pointer: %T", out)
+	}
+	outVal = outVal.Elem()
+	if outVal.Kind() != reflect.Struct {
+		return fmt.Errorf("discovery: can only unmarshal into a struct pointer: %T", out)
+	}
+	outTyp := outVal.Type()
+
+	cfgTyp := getConfigType(outTyp)
+	cfgPtr := reflect.New(cfgTyp)
+	cfgVal := cfgPtr.Elem()
+
+	// Copy shared fields (defaults) to dynamic value.
+	var configs *Configs
+	for i, n := 0, outVal.NumField(); i < n; i++ {
+		if outTyp.Field(i).Type == configsType {
+			configs = outVal.Field(i).Addr().Interface().(*Configs)
+			continue
+		}
+		if cfgTyp.Field(i).PkgPath != "" {
+			continue // Field is unexported: ignore.
+		}
+		cfgVal.Field(i).Set(outVal.Field(i))
+	}
+	if configs == nil {
+		return fmt.Errorf("discovery: Configs field not found in type: %T", out)
+	}
+
+	// Unmarshal into dynamic value.
+	if err := unmarshal(cfgPtr.Interface()); err != nil {
+		return replaceYAMLTypeError(err, cfgTyp, outTyp)
+	}
+
+	// Copy shared fields from dynamic value.
+	for i, n := 0, outVal.NumField(); i < n; i++ {
+		if cfgTyp.Field(i).PkgPath != "" {
+			continue // Field is unexported: ignore.
+		}
+		outVal.Field(i).Set(cfgVal.Field(i))
+	}
+
+	var err error
+	*configs, err = readConfigs(cfgVal, outVal.NumField())
+	return err
+}
+
+func readConfigs(structVal reflect.Value, startField int) (Configs, error) {
+	var (
+		configs Configs
+		targets []*targetgroup.Group
+	)
+	for i, n := startField, structVal.NumField(); i < n; i++ {
+		field := structVal.Field(i)
+		if field.Kind() != reflect.Slice {
+			panic("discovery: internal error: field is not a slice")
+		}
+		for k := 0; k < field.Len(); k++ {
+			val := field.Index(k)
+			if val.IsZero() || (val.Kind() == reflect.Ptr && val.Elem().IsZero()) {
+				key := configFieldNames[field.Type().Elem()]
+				key = strings.TrimPrefix(key, configFieldPrefix)
+				return nil, fmt.Errorf("empty or null section in %s", key)
+			}
+			switch c := val.Interface().(type) {
+			case *targetgroup.Group:
+				// Add index to the static config target groups for unique identification
+				// within scrape pool.
+				c.Source = strconv.Itoa(len(targets))
+				// Coalesce multiple static configs into a single static config.
+				targets = append(targets, c)
+			case Config:
+				configs = append(configs, c)
+			default:
+				panic("discovery: internal error: slice element is not a Config")
+			}
+		}
+	}
+	if len(targets) > 0 {
+		configs = append(configs, StaticConfig(targets))
+	}
+	return configs, nil
+}
+
+// MarshalYAMLWithInlineConfigs helps implement yaml.Marshal for structs
+// that have a Configs field that should be inlined.
+func MarshalYAMLWithInlineConfigs(in interface{}) (interface{}, error) {
+	// This function can be removed if https://github.com/go-yaml/yaml/issues/642 is fixed.
```

Comment:
Is that true for v2 too?

Reply:
Yep

Reply:
https://play.golang.com/p/DjhRLsGmbza

Reply:
Actually... Maybe not... doing some tests.

Reply:
It's complicated. The behavior is different between the two versions.

But before getting into that issue, both v2 and v3 only allow inlining structs (or maps, but inlined maps destroy strictness and are another can of worms), and we have a slice of configs that we would like to inline. This is a separate issue and it makes that comment invalid because https://github.com/go-yaml/yaml/issues/642 isn't the only blocker.

Back to the original issue, in v3 it seems that when a struct implementing `yaml.Unmarshaler` is inlined in a parent struct, the fields that may or may not be in the inlined struct are not taken into account when enforcing strictness on the parent struct. So fields that exist in the inlined struct or exist in its custom unmarshaling logic will cause "field not found" errors on the parent.

In v2, it seems to ignore any custom unmarshaling logic when a `yaml.Unmarshaler` is inlined. Unmarshal is never called on the inlined struct. Strictness is enforced based on the union of fields in the parent struct and inlined struct. Again, fields that only exist in custom unmarshaling logic (which is never called) will cause "field not found" errors on the parent.

Reply:
Here's a demo for v2: https://play.golang.com/p/yvcjhettRth

Reply:
I removed the comment.